django/db/models/lazy.py

```diff
@@ -0,0 +1,63 @@
+from contextlib import contextmanager
+
+from asgiref.local import Local
+
+
+def FETCH_ONE(instance, field, fetch_for_instances, **kwargs):
+    fetch_for_instances((instance,))
+
+
+def FETCH_PEERS(instance, field, fetch_for_instances, **kwargs):
+    instances = [
+        peer
+        for weakref_peer in instance._state.peers
+        if (peer := weakref_peer()) is not None
+    ]
+    if not instances:
+        # Peers aren’t tracked for QuerySets returning a single instance
+        instances = (instance,)
+
+    fetch_for_instances(instances)
+
+
+class LazyFieldAccess(Exception):
+    """Blocked lazy access of a model field."""
+
+    pass
+
+
+def RAISE(instance, field, fetch_for_instances, **kwargs):
```

Comment:
I was thinking of that, but then maybe users would want “fetch peers” + warnings, or log messages directly rather than warnings. Maybe something we can instead document the pattern for? Like:

```python
def warn_then_fetch_peers(...):
    logger.warning("fetching peers!")
    return FETCH_PEERS(...)

set_default_lazy_mode(warn_then_fetch_peers)
```

Reply:
I see, that makes sense to separate non-terminal _surfacing_ of _leaks_ from the actual fetching implementation. I wonder if there is anything that we could provide to avoid such boiler plate though.